hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/ResourceManager.java

```diff
@@ -824,25 +824,29 @@ public void handle(RMFatalEvent event) {
    * Transition to standby in a new thread.
    */
   public void handleTransitionToStandByInNewThread() {
-    new Thread() {
-      @Override
-      public void run() {
-        if (rmContext.isHAEnabled()) {
-          try {
-            // Transition to standby and reinit active services
-            LOG.info("Transitioning RM to Standby mode");
-            transitionToStandby(true);
-            EmbeddedElector elector = rmContext.getLeaderElectorService();
-            if (elector != null) {
-              elector.rejoinElection();
-            }
-          } catch (Exception e) {
-            LOG.fatal("Failed to transition RM to Standby mode.", e);
-            ExitUtil.terminate(1, e);
+    Thread standByTransitionThread = new Thread(new StandByTransitionThread());
```

Comment:
Sorry for not identifying this earlier. We should make this thread-safe in case this is triggered by two critical threads failing at the same time.